WordPress/gutenberg
# Validate and protect corrupted block content

Closes #1735
Closes #1736
Related: #391

This pull request seeks to validate blocks parsed from post content. It does so by regenerating the expected save output upon the initial parse and comparing it against the actual content. If there is a mismatch, the block is flagged as invalid and the user is barred from editing the block (aside from moving or deleting) with a visual indicator of the invalid state:

![Invalid block](https://user-images.githubusercontent.com/1779930/28286942-e264ca1a-6b07-11e7-985b-6586199d0a6f.png)

__Implementation notes:__

- Yes, there's likely a performance impact of creating an initial serialization of the block
- Yes, there's likely false positives that can occur, particular on order of attributes or class names (which in reality should not have an impact on behavior of content). To reduce false positives, both contents are run through the same HTML beautifier we use for serialization. False positives from reordering attributes seems acceptable to me, as we should discourage manual markup modifications, or at least transition the user to a "manual mode". Enhancements could be made here. Editing block markup in Text mode should be unaffected.

__Testing instructions:__

I've left the initial Cover Image block in Demo content as invalid, because I'd like to revisit whether we really want [nested `section` tags](https://github.com/WordPress/gutenberg/blob/6a8e2485675b20b871f9178b51f116176347eab3/blocks/library/cover-image/index.js#L157-L158) in the serialized output. Verify that the first block in post content is flagged as invalid and cannot be edited.

1. Navigate to Gutenberg > Demo
2. Note that the first block has a "Corrupt" warning like the one shown above
